### PR TITLE
Throw an AuthenticationException for 401 status codes

### LINKED ISSUE #153: Throw more detailed exceptions
We should add these exceptions to make it quicker, cleaner and easier for devs using the SDK:

 * `TokenExpiredException` - indicates the existing token being used is expired and needs to be refreshed
 * `AuthenticationException` - indicates the SDK was unable to successfully authenticate with the API

### PR DESCRIPTION
**Pull request by anthonyterrell**
Closes #153 

**⚠️ This should be merged in before #214 to prevent merge conflicts** 

This PR uses the existing `AuthenticationException` when the status code `401` is returned.

### CODE REVIEW

File: tests/Authentication/AuthenticationIntegrationTest.php [hunk 1]
```diff
@@ -241,4 +242,13 @@ public function testFetchTokensWithNullCredentialsThrowsException()
         $authenticator = new Authenticator($this->guzzle);
         $authenticator->getAuthHeader();
     }
+
+    public function testAuthenticationExceptionThrown()
```
Comment (anthonyterrell) [inline tests/Authentication/AuthenticationIntegrationTest.php]: I don't _love_ this test name and am open to suggestion :)

File: src/Http/Handlers/ClientErrorHandler.php [hunk 2]
```diff
@@ -15,7 +15,7 @@ public function __invoke(callable $handler)
         return function (RequestInterface $request, array $options = []) use ($handler) {
             return $handler($request, $options)->then(
                 function (ResponseInterface $response) use ($request) {
-                    if ($response->getStatusCode() >= 400) {
+                    if ($response->getStatusCode() >= 400 && $response->getStatusCode() !== 401) {
```
Comment (bkuhl) [inline src/Http/Handlers/ClientErrorHandler.php]: It'd probably be a good idea to add a test case for this handler to ensure it ignores response code 401

File: src/Http/Handlers/AuthenticationHandler.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+class AuthenticationHandler
+{
+    public function __invoke(callable $handler)
+    {
+        return function (RequestInterface $request, array $options = []) use ($handler) {
+            return $handler($request, $options)->then(
+                function (ResponseInterface $response) use ($request) {
+                    if ($response->getStatusCode() === 401) {
+                        throw new AuthenticationException('Authentication Error', $request, $response);
```
Comment (bkuhl) [inline src/Http/Handlers/AuthenticationHandler.php]: What do you about changing the exception message to be `Invalid credentials or expired refresh token`

File: tests/Authentication/AuthenticationIntegrationTest.php (hunk 1 above)
Comment (bkuhl) [reply tests/Authentication/AuthenticationIntegrationTest.php]: `testAuthorizationExceptionWhenUnauthorizedStatusCode`?  I tend to lean towards longer test names to try to describe what happens and when.

File: src/Http/Handlers/ClientErrorHandler.php (hunk 2 above)
Comment (anthonyterrell) [reply src/Http/Handlers/ClientErrorHandler.php]: Great idea.

File: src/Http/Handlers/AuthenticationHandler.php (hunk 3 above)
Comment (anthonyterrell) [reply src/Http/Handlers/AuthenticationHandler.php]: I'm down to change this, although if both are possible returns it feels like the message should be inclusive to both. What about `Not Authorized`, as we use that verbiage in the [docs](https://developer.helpscout.com/mailbox-api/overview/status_codes/)?

Comment (bkuhl) [reply src/Http/Handlers/AuthenticationHandler.php]: I like the idea of providing more guidance when/where we can, but am not opposed to staying generic since there are edge cases we may not be aware of.

File: tests/Http/Handlers/ClientHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public function testRequestExceptionThrownWhenAccessDenied()
+    {
+        $this->expectException(RequestException::class);
+
+        $this->stubResponse($this->getResponse(403, json_encode([])));
+
+        $this->client->customers()->get(1);
+    }
```
Comment (bkuhl) [inline tests/Http/Handlers/ClientHandlerTest.php]: I think we just add 1 test here that doesn't throw an exception if the response is `401` and it'll be good to go

Comment (anthonyterrell) [reply tests/Http/Handlers/ClientHandlerTest.php]: While digging in and writing this test I found out that this testable scenario will only occur if **the Authenciationhandler is _not_ added to the stack** [here](https://github.com/helpscout/helpscout-api-php/blob/27395a069c02a01f35957b34deb5175112746bda/src/Http/RestClientBuilder.php#L94). If I'm correct, as long as the AuthenticationHandler is pushed to the Handler stack, then a `401` would return an exception and not the next request. 

Example test:

```php
public function testRequestExceptionNotThrownWhenNotAuthorized()
    {
        $this->mockHandler = new MockHandler();
        $handler = HandlerStack::create($this->mockHandler);

        $handler->push(Middleware::history($this->history));
        // AuthenticationHandler not pushed to Handler stack
        $handler->push(new ClientErrorHandler());
        $handler->push(new RateLimitHandler());
        $handler->push(new ValidationHandler());

        $client = new Client(['handler' => $handler, 'http_errors' => false]);
        $this->authenticator = new Authenticator($client);
        $this->authenticator->setAccessToken('abc123');

        $this->client = new ApiClient(
            new RestClient($client, $this->authenticator)
        );
        $this->stubResponse($this->getResponse(401, json_encode([])));
        $customer = $this->client->customers()->get(1);
        $this->assertInstanceOf(Customer::class, $customer);
    }
```

Comment (bkuhl) [reply tests/Http/Handlers/ClientHandlerTest.php]: We could reduce the complexity of the test by only testing the handler itself, avoiding us having to build the several `Client` and `Handler` objects.  I haven't executed this so this isn't a complete test, but here's an example of what I'm thinking:

```
$expectedResponse = Mockery::mock(ResponseInterface::class);
$handler = new AuthenticationHandler();
$response = $handler(function () use ($expectedResponse) {
    $promise = new Promise(function () use (&$promise, $expectedResponse) {
        $promise->resolve(expectedResponse);
    });
    return $promise;
});
$this->assertEquals($expectedResponse, $response);
```

There's docs on Guzzle's promises at https://github.com/guzzle/promises.

Comment (anthonyterrell) [reply tests/Http/Handlers/ClientHandlerTest.php]: Very nice, 👏

Comment (anthonyterrell) [reply tests/Http/Handlers/ClientHandlerTest.php]: I was fussing with this for a bit, but was able to trim down this test and still have it passing. Let me know what you think.

```php
public function testRequestExceptionNotThrownWhenNotAuthorized()
    {
        $this->mockHandler = new MockHandler();
        $handler           = HandlerStack::create($this->mockHandler);

        $client              = new Client(['handler' => $handler]);
        $this->authenticator = new Authenticator($client);
        $this->authenticator->setAccessToken('abc123');

        $this->client = new ApiClient(
            new RestClient($client, $this->authenticator)
        );

        $this->stubResponse($this->getResponse(401, json_encode([])));

        $this->assertInstanceOf(
            Customer::class,
            $this->client->customers()->get(1)
        );
    }
```

Comment (bkuhl) [reply tests/Http/Handlers/ClientHandlerTest.php]: Looks great, well done!
